### PR TITLE
Added D* Search to path planning folder

### PR DESCRIPTION
**Pull request by nirnayroy**

### CODE REVIEW

**Comment by nirnayroy** [comment]
this PR implements the D* search algorithm

File: PathPlanning/DStar/dstar.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
+import math
```
Comment (AtsushiSakai) [inline PathPlanning/DStar/dstar.py:1]: Could you add header comment which includes short description, author name, and reference paper?

File: PathPlanning/DStar/dstar.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+from sys import maxsize
+
+import matplotlib.pyplot as plt
+
+show_animation = True
+
+
+class State(object):
```
Comment (AtsushiSakai) [inline PathPlanning/DStar/dstar.py:10]: This project only support python3, so I think no need to inherit `object`.

File: PathPlanning/DStar/dstar.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        e: oparent of current state
+        *: closed state
+        s: current state
+        '''
+        if state not in ["s", ".", "#", "e", "*"]:
+            return
+        self.state = state
+
+
+class Map(object):
```
Comment (AtsushiSakai) [inline PathPlanning/DStar/dstar.py:41]: This project only support python3, so I think no need to inherit `object`.

File: PathPlanning/DStar/dstar.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+    def cost(self, state):
+        if self.state == "#" or state.state == "#":
+            return maxsize
+
+        return math.sqrt(math.pow((self.x - state.x), 2) +
+                         math.pow((self.y - state.y), 2))
+
+    def set_state(self, state):
+        '''
```
Comment (AtsushiSakai) [inline PathPlanning/DStar/dstar.py:29]: Could you please use double quote for docstring?

File: PathPlanning/DStar/dstar.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+    def set_obstacle(self, point_list):
+        for x, y in point_list:
+            if x < 0 or x >= self.row or y < 0 or y >= self.col:
+                continue
+
+            self.map[x][y].set_state("#")
+
+
+class Dstar(object):
```
Comment (AtsushiSakai) [inline PathPlanning/DStar/dstar.py:78]: This project only support python3, so I think no need to inherit `object`.

File: README.md [hunk 6]
```diff
@@ -301,6 +302,19 @@ In the animation, cyan points are searched nodes.
 
 Its heuristic is 2D Euclid distance.
 
+### D\* algorithm
+
+This is a 2D grid based the shortest path planning with D star algorithm.
+
+![figure at master · nirnayroy/intelligentrobotics](https://github.com/nirnayroy/intelligent-robotics/blob/main/dstar.gif)
```
Comment (AtsushiSakai) [inline README.md:309]: [comment] I will change the link to my gif repository.

File: docs/modules/path_planning.rst [hunk 7]
```diff
@@ -455,4 +472,4 @@ Ref:
 .. |dubins| image:: https://github.com/AtsushiSakai/PythonRoboticsGifs/raw/master/PathPlanning/DubinsPath/animation.gif?raw=True
 .. |RSPlanning| image:: https://github.com/AtsushiSakai/PythonRoboticsGifs/raw/master/PathPlanning/ReedsSheppPath/animation.gif?raw=true
 .. |RSPlanning2| image:: https://github.com/AtsushiSakai/PythonRoboticsGifs/raw/master/PathPlanning/LQRPlanner/animation.gif?raw=true
-.. |15| image:: https://github.com/AtsushiSakai/PythonRoboticsGifs/raw/master/PathPlanning/FrenetOptimalTrajectory/animation.gif
+.. |15| image:: https://github.com/AtsushiSakai/PythonRoboticsGifs/raw/master/PathPlanning/FrenetOptimalTrajectory/animation.gif
```
Comment (AtsushiSakai) [inline docs/modules/path_planning.rst:475]: Why diff is here?

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for great PR!!. I have some comments. PTAL.

File: README.md (hunk 6 above)
Comment (nirnayroy) [reply README.md:309]: Should I create a pull request to your gifs repository?

Comment (AtsushiSakai) [reply README.md:309]: Thanks. You don't need to do that. I can do that.

File: PathPlanning/DStar/dstar.py [hunk 8]
```diff
@@ -189,17 +198,31 @@ def modify(self, state):
...
+        ox.append(40.0)
+        oy.append(60.0 - i)
+    m.set_obstacle([[i, j] for i, j in zip(ox, oy)])
 
     start = [1, 2]
     goal = [17, 19]
     if show_animation:
-        plt.plot([x for (x, y) in obstacle_list],
-                 [y for (x, y) in obstacle_list], ".k")
+        plt.plot([ox, oy, ".k"])
```
Comment (AtsushiSakai) [inline PathPlanning/DStar/dstar.py:225]: I think this should be `plt.plot(ox, oy, ".k")`? In my environment, simulation has an error: 

```
/usr/local/lib/python3.9/site-packages/numpy/core/_asarray.py:171: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  return array(a, dtype, copy=False, order=order, subok=True)
TypeError: float() argument must be a string or a number, not 'list'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/atsushisakai/PycharmProjects/PythonRobotics/PathPlanning/DStar/dstar.py", line 240, in <module>
    main()
  File "/Users/atsushisakai/PycharmProjects/PythonRobotics/PathPlanning/DStar/dstar.py", line 225, in main
    plt.plot([ox, oy, ".k"])
  File "/usr/local/lib/python3.9/site-packages/matplotlib/pyplot.py", line 2988, in plot
    return gca().plot(
  File "/usr/local/lib/python3.9/site-packages/matplotlib/axes/_axes.py", line 1607, in plot
    self.add_line(line)
  File "/usr/local/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 2105, in add_line
    self._update_line_limits(line)
  File "/usr/local/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 2127, in _update_line_limits
    path = line.get_path()
  File "/usr/local/lib/python3.9/site-packages/matplotlib/lines.py", line 1022, in get_path
    self.recache()
  File "/usr/local/lib/python3.9/site-packages/matplotlib/lines.py", line 668, in recache
    y = _to_unmasked_float_array(yconv).ravel()
  File "/usr/local/lib/python3.9/site-packages/matplotlib/cbook/__init__.py", line 1333, in _to_unmasked_float_array
    return np.asarray(x, float)
  File "/usr/local/lib/python3.9/site-packages/numpy/core/_asarray.py", line 102, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: setting an array element with a sequence.
```

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Sorry for late reply. Could you please fix the issue?

File: PathPlanning/DStar/dstar.py (hunk 8 above)
Comment (nirnayroy) [reply PathPlanning/DStar/dstar.py]: you are right...I have fixed it. Sorry for delay.

**Review by AtsushiSakai — APPROVED** [review]
LGTM. When CI is green. I will merge it. Thanks for your contributions!! @nirnayroy

**Pull request merged**